ublox_derive/src/output.rs

```diff
@@ -0,0 +1,572 @@
+                let bytes = self.#name.into_raw().to_le_bytes()
+            });
+        } else if !f.is_field_raw_ty_byte_array() {
+            pack_fields.push(quote! {
+                let bytes = self.#name.to_le_bytes()
+            });
+        } else {
+            pack_fields.push(quote! {
+                let bytes: &[u8] = &self.#name;
+            });
+        }
+
+        write_fields.push(pack_fields.last().unwrap().clone());
+        write_fields.push(quote! {
+            out.write(&bytes)?;
+            checksum_calc.update(&bytes)
+        });
+        for i in 0..size_bytes.get() {
+            let byte_off = off.checked_add(i).unwrap();
+            pack_fields.push(quote! {
+                ret[#byte_off] = bytes[#i]
+            });
+        }
+
+        off += size_bytes.get();
+    }
+    let builder_attr = if pack_descr
+        .header
+        .flags
+        .iter()
+        .any(|x| *x == PacketFlag::DefaultForBuilder)
+    {
+        quote! { #[derive(Default)] }
+    } else {
+        quote! {}
+    };
+    let mut ret = Vec::with_capacity(4);
+    let struct_comment = &pack_descr.comment;
+    ret.push(quote! {
+        #[doc = #struct_comment]
+        #[doc = "Struct that used as \"builder\" for packet"]
+        #builder_attr
+        pub struct #payload_struct {
+            #(#fields),*
+        }
+    });
+
+    if let Some(packet_payload_size) = pack_descr.packet_payload_size() {
+        let packet_size = packet_payload_size + 8;
+        let packet_payload_size_u16 = u16::try_from(packet_payload_size).unwrap();
+        ret.push(quote! {
+            impl #payload_struct {
+                pub const PACKET_LEN: usize = #packet_size;
+
+                #[inline]
+                pub fn into_packet_bytes(self) -> [u8; Self::PACKET_LEN] {
+                    let mut ret = [0u8; Self::PACKET_LEN];
+                    ret[0] = SYNC_CHAR_1;
+                    ret[1] = SYNC_CHAR_2;
+                    ret[2] = #main_name::CLASS;
+                    ret[3] = #main_name::ID;
+                    let pack_len_bytes = #packet_payload_size_u16 .to_le_bytes();
+                    ret[4] = pack_len_bytes[0];
+                    ret[5] = pack_len_bytes[1];
+                    #(#pack_fields);*;
+                    let (ck_a, ck_b) = ubx_checksum(&ret[2..#packet_size-2]);
+                    ret[#packet_size-2] = ck_a;
+                    ret[#packet_size-1] = ck_b;
+                    ret
+                }
+            }
+            impl From<#payload_struct> for [u8; #packet_size] {
+                fn from(x: #payload_struct) -> Self {
+                    x.into_packet_bytes()
+                }
+            }
+        });
+
+        ret.push(quote! {
+            impl UbxPacketCreator for #payload_struct {
+                #[inline]
+                fn create_packet<T: MemWriter>(self, out: &mut T) -> Result<(), MemWriterError<T::Error>> {
+                    out.reserve_allocate(#packet_size)?;
```

Comment:
Is the only reason we want `reserve_allocate` for performance reasons?

It occurred to me just now that it'd be weird to write `reserve_allocate` for, for example, a SpiWriter that writes directly to a spi port.

Reply:
> SpiWriter that writes directly to a spi port

I don't think that it is possible. API is designed for fill some buffer with data. For buffer reuse, for writing to some special location like DAM buffer continiously mapped to physical memory  and so on. That's why it called `MemWriter`. 

It can try to use this API to write directly to device, but in this case it would be hard to
handle errors like "try again latter" (EGAIN/EBUSY), because of you have not buffer, but you need restart from specific location. This is API not for direct device access.

Reply:
For sure, there are some devices which would be very hard to interface with. There is a [blocking SPI bus](https://github.com/rust-embedded/embedded-hal/blob/master/src/blocking/spi.rs) in the embedded hal, though, which would probably work fine.

Reply:
Obviously any unblocking API can be used in blocking way via busy loop. But again why care? 

This is API designed for filling memory buffers, not for "direct" transfer. If somebody not care about performance, and don't use DMA, so his/her CPU slow-down and starts work with SPI rate, like 10-20Mhz, then obviously that person doesn't care about performance at all, why care about that case?

Plus `reserve_allocate` is not only for performance reason, it is also for handling "out-of-memory" errors like https://doc.rust-lang.org/beta/std/vec/struct.Vec.html#method.try_reserve .

Plus what is the gain here? I see only the reason to use "direct" transfer in compare with into_packet_bytes/DmaBuf::write in increasing of code size by decreasing stack size. So basically if you use "direct" transfer you increase code size (many writes), if you use "into_packet_bytes"/DmaBuf::write you decrease code size, but increase stack or heap usage. 

And if we talk about our case, then our stack/heap usage is really small. May be this is significant for really small microcontrollers with 2K SRAM (stack/heap) and 32K NOR flash (code), but I really doubt that there is sense to use such kind of microcontrollers with GPS.

Reply:
> why care about that case?

Ease of implementation for the end user. And of course, for the memory-sensitive - you say there's no sense in using a small MCU with a GPS module, but arduino's project hub has [several projects](https://create.arduino.cc/projecthub/projects/tags/gps) which combine GPS modules with controllers that have <2KB of ram.

> Plus reserve_allocate is not only for performance reason, it is also for handling "out-of-memory" errors

out-of-memory errors I think could be handled by failing in the write() call as well.

Either way, I'm fine with leaving it as it is, I was just thinking about the ramifications of this particular API.

Reply:
> out-of-memory errors I think could be handled by failing in the write() call as well

In some cases no. For example `alloc::vec::Vec` it would be troublesome.  Only `Vec::try_reserve` return out-of-memory error. `Write::write` works without errors.

> but arduino's project hub 

In this case `PacketBuilder::into_packet_bytes` in form of `const fn` function would be much nicer. Of course when `rustc` allow `for` in `const fn` for `ubx_checksum`. Then instead of load chunks of data into SRAM for sending, it would be possible to send one whole packet in one call, and flash to hold const byte array with packet.